### PR TITLE
Questions

### PR DESCRIPTION
**Pull request by AnikaC-git**
## Reason for changes and summary
During code review of consultation request processing it transpired that the question handling as was implemented was not ideal by means that for each data type that has questions attached to it, a new table and likely also audit table would have to be created for the questions. Based on an in person discussion, it was decided to hold all questions (independent of data type they relate to) in one single table and a link table to allow for searches across all questions without having to join tables.

## Checklist for submitter & reviewer

- [x] copy constructor copies the primary key Id of the entity
- [x] copy constructor copies all other fields of the entity, starting with `super(other)`
- [x] unique constraints added if appropriate
- [x] text definition added to columns which could have more than 255 characters
- [x] foreign key names are in the form `<entityClass>Id`, and no typos
- [x] during run, glowroot is checked for query speed and indexes added if needed
- [x] validation run completed and validated

### CODE REVIEW

**Review by klapaukh — COMMENTED** [review]
So I think I'm maybe a bit confused by what is meant be going on.

Some questions:
 - Why is there both `RequestAnswer` & `ConsultationRequestQuestion`? They seem like they store the same data?
 - In `RequestAnswer` how do you know which table `parentId` links to? 
 - Why is `parentId` called that? How is it the parent?
 - How do we ensure that we never get into a state where the `RequestAnswer` table has broken links in it?

**Comment by AnikaC-git** [comment]
@klapaukh 

Sorry, only commented on this in Emap-Core. The reason why there are "duplicated" tables is because I did not delete the old classes yet, mainly because I did not know how crude we are about the deletions when we do such changes. If you'd be happy to confirm that for me, I'll happily delete all the old classes.

As for the parent id, this is effectively a link to the entity that the question relates to (ie. id for either lab sample or consult request that triggered the generation), but as you point out correctly there is no link back to the parent table. Would you want this to be an ENUM indicating which data type generated the question? Or is there a way to get a reference to the table itself?

What do you mean by broken links?

**Comment by klapaukh** [comment]
- The code should be clean. So unused stuff should be deleted. 
- I'm not aware of any useful way to create a reference to the table, apart from using something like an ENUM.
- Normally a foreign key relationship is used to ensure that the  ids exist in both tables, i.e. that the request that the question is linked to exists, and that if it is deleted, the question is too. But since you don't have a foreign key relationship it's possible to have a dangling reference. A value in `parentId` that doesn't have a matching entry in the parent table.

**Comment by AnikaC-git** [comment]
I've now deleted all the superfluous classes and adding in the parent string. Did you want me to deal with the "dangling references" somehow or do we accept that this is something that happens?

**Comment by stefpiatek** [comment]
currently we don't have a message that causes a deletion of a question (they're linked to lab samples which aren't removed and consult requests which also aren't removed) so I think we could probably get away with documenting that the questions are currently only cumulative? 

For reference to table, could we get this from the class name of the parent entity or maybe there's a hibernate method to get the table name and pass that in at the same time?

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/Question.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import java.time.Instant;
+
+/**
+ * Questions that can be attached to several data types, such as lab samples or consultation requests. Independent of
+ * which type of question, these are all held together in one table and reference back to the entity they relate too.
```
Comment (stefpiatek) [inline inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/Question.java:18]:
```suggestion
 * which type of question, these are all held together in one table and reference back to the entity they relate to.
```

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/Question.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+@AuditTable
+public class Question extends TemporalCore<Question, QuestionAudit> {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private long questionId;
+
+    @Column(columnDefinition = "text", nullable = false)
+    private String question;
+
+    public Question() {}
```
Comment (stefpiatek) [inline inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/Question.java:36]: FYI, can just use the lombok NoArgsConstructor annotation

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/RequestAnswer.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.JoinColumn;
+import javax.persistence.ManyToOne;
+import java.time.Instant;
+
+/**
+ * Questions generated from specific data types (e.g. lab samples or consultation requests) trigger answers that or
+ * held as RequestAnswers. RequestAnswers therefore not only hold the answer as such but also link the answer to the
```
Comment (stefpiatek) [inline inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/RequestAnswer.java:20]: This feels a bit confusing to read, I think there's a mistake in `tiggers answers that or held as...`, probably worth revisiting

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/RequestAnswer.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+*/
+@Entity
+@Data
+@EqualsAndHashCode(callSuper = true)
+@ToString(callSuper = true)
+@AuditTable
+public class RequestAnswer extends TemporalCore<RequestAnswer, RequestAnswerAudit> {
+    @Id
+    @GeneratedValue(strategy = GenerationType.AUTO)
+    private long answerId;
```
Comment (stefpiatek) [inline inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/RequestAnswer.java:32]: entity id should always be the entity name + Id

```suggestion
    private long requestAnswerId;
```

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/RequestAnswer.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    public RequestAnswer() {}
+
+    /**
+     * Minimal request answer constructor that requires that the answer and the timestamps for when EMAP
+     * started processing this data type and from which time point the question information is valid from.
+     * @param questionId    Question this answer relates to
+     * @param answer        The actual question string linked to a data type
+     * @param parentTable   Data type that triggered question-answer pair creation, e.g. lab sample or consultation
+     *                      request
+     * @param parentId      Entity (e.g. lab sample or consultation request) that triggered the
```
Comment (stefpiatek) [inline inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/RequestAnswer.java:56]: looks like we're missing the end here!
```suggestion
     * @param parentId      Entity (e.g. lab sample or consultation request) that triggered the
```

**Review by stefpiatek — COMMENTED** [review]
Looks good, seems like a pretty clean way to solve the problem. Spotted some documentation changes and some field naming changes that we'd need.

**Comment by klapaukh** [comment]
> Did you want me to deal with the "dangling references" somehow or do we accept that this is something that happens?

Data integrity is important! If we can provide some kind of assurance that it won't go wrong we should.

File: inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/RequestAnswer.java (hunk 3 above)
Comment (AnikaC-git) [reply inform-db/src/main/java/uk/ac/ucl/rits/inform/informdb/questions/RequestAnswer.java:20]: Had another attempt which hopefully is clearer.

**Comment by AnikaC-git** [comment]
Addressed all the comments as was available. Might be ready for a validation run once the core branch is ready too which I'm going to look at now.

**Review by stefpiatek — APPROVED** [review]
Nicely done, I think the only thing we haven't addressed is having some sort of constraint to ensure data integrity - I know @klapaukh had some ideas for this

**Comment by stefpiatek** [comment]
it's not a huge slow down but probaly worth putting an index on `parent_table_id` 
image

**Comment by stefpiatek** [comment]
> it's not a huge slow down but probably putting an index on `parent_table_id` because it'll only get worse as the number of answers grows
> image

I'm not sure if this index was added?

**Comment by AnikaC-git** [comment]
I'll double-check before merging. Somehow I thought I did ... At the moment some trouble with EMAP core anyway ... :/

**Comment by AnikaC-git** [comment]
Missing index added and test still work locally as before. Speed improvements are not visible when running locally.

**Pull request merged**